### PR TITLE
PF-1130: Make version lookup configurable.

### PR DESCRIPTION
**Pull request by marikomedlock**
Added a `VersionScript` that is a new optional property of a `ServerSpecification`. If defined, it will run after deployment, but before the test scripts. In this way, it is very similar to a `DeploymentScript`, which is another optional server property.

Each `VersionScript` sub-class overrides the `determineVersion` method that contains the implementation of how to lookup the version. I included a skeleton sub-class as an example.

The `determineVersion` method returns an instance of the `VersionScriptResult` class, which includes properties that specify the server's version. We will likely want to add to this class to include other services and deployment types. In the future, we way want to sub-class `VersionScriptResult` to accommodate deployments that are very different (e.g. Broad vs Verily), but for now I just defined one class for simplicity. The main `TestRunner` class writes out the `VersionScriptResult` to a new output file.

Moved the Kubernetes version lookup into a `VersionScript` sub-class also, since that's a kind of version information. And since it's not going to apply to every environment (e.g. Verily deployments, Broad personal environments), it makes sense to enable it per `ServerSpecification`.

### CODE REVIEW

**Comment by ichengchang** [comment]
@marikomedlock I updated the `ReadFromTerraHelmfileRepo` with steps to parse `terra-helmfile` values. Bumped version to 0.0.10. 

A workflow run with `VersionScriptSpecification` config is [here](https://github.com/DataBiosphere/terra-workspace-manager/runs/4155383138?check_suite_focus=true).

Let's merge this PR first before the others.

File: src/main/java/bio/terra/testrunner/common/HelmRelease.java
Comment (marikomedlock) [inline src/main/java/bio/terra/testrunner/common/HelmRelease.java:5]: nit: could you add a description here? I think this is going to have one version per service right? if so, that information, or whatever you expect the map keys to be would be helpful.

File: src/main/java/bio/terra/testrunner/common/HelmReleaseVersion.java
Comment (marikomedlock) [inline src/main/java/bio/terra/testrunner/common/HelmReleaseVersion.java:5]: nit: same as above. could you please add a description. maybe with a link to the helm repo where these things are getting pulled from?

File: src/main/java/scripts/versionscripts/ReadFromTerraHelmfileRepo.java
Comment (marikomedlock) [inline src/main/java/scripts/versionscripts/ReadFromTerraHelmfileRepo.java:49]: If this doesn't work now, I'd prefer to leave it commented out with a TODO (JIRA ticket) to reenable when devops is ready. We can probably comment it out of the `VersionScriptResult` class also with the same TODO.

File: src/main/java/scripts/versionscripts/ReadFromTerraHelmfileRepo.java
Comment (marikomedlock) [inline src/main/java/scripts/versionscripts/ReadFromTerraHelmfileRepo.java:62]: nit: I think this comment (at least the part about pulling from env vars or terra-helmfile) no longer applies.

File: src/main/java/scripts/versionscripts/ReadFromTerraHelmfileRepo.java
Comment (marikomedlock) [inline src/main/java/scripts/versionscripts/ReadFromTerraHelmfileRepo.java:75]: I think this helper function is looping through the override helm values and replacing any base helm values, then returning the now-modified base helm values map -- is that right? A brief explanation here in a comment would be great.

File: src/main/java/bio/terra/testrunner/common/HelmRelease.java
Comment (ichengchang) [reply src/main/java/bio/terra/testrunner/common/HelmRelease.java:9]: Added description to `HelmRelease`.

File: src/main/java/bio/terra/testrunner/common/HelmReleaseVersion.java
Comment (ichengchang) [reply src/main/java/bio/terra/testrunner/common/HelmReleaseVersion.java:11]: Added description to `HelmReleaseVersion`.

File: src/main/java/scripts/versionscripts/ReadFromTerraHelmfileRepo.java
Comment (ichengchang) [reply src/main/java/scripts/versionscripts/ReadFromTerraHelmfileRepo.java]: Commented out `importComponentVersions` pending `DevOps` readiness. Will track this with Jira ticket [QA-1643](https://broadworkbench.atlassian.net/browse/QA-1643).

File: src/main/java/scripts/versionscripts/ReadFromTerraHelmfileRepo.java
Comment (ichengchang) [reply src/main/java/scripts/versionscripts/ReadFromTerraHelmfileRepo.java]: Updated comments, removed reference to `envvars`.

File: src/main/java/scripts/versionscripts/ReadFromTerraHelmfileRepo.java
Comment (ichengchang) [reply src/main/java/scripts/versionscripts/ReadFromTerraHelmfileRepo.java:81]: Added comments for the `merge` method.

**Comment by ichengchang** [comment]
@marikomedlock In my latest changes. I moved the `Helm-*` classes into a new package `scripts.versionscripts.model` and refactored `HelmReleaseVersion` as inner class of `HelmRelease`. You can merge this PR if everything looks good. After you merge this PR, I'll resolve any potential conflict for [PR#21](https://github.com/DataBiosphere/terra-test-runner/pull/21) and update the code to use `VersionScript`.

**Review by ichengchang — APPROVED** [review]
LGTM

**Pull request merged**
